In [1]:
import pandas as pd
import numpy as np

<img src="./ex1.JPG">

First of all I import csv files extracted from the excel sheet.

In [2]:
df_cc = pd.read_csv("./SONY_Exercise_1 - Customer claim.csv", index_col = 'CustomerNR' )
df_sap = pd.read_csv("./SONY_Exercise_1 - SAP.csv",index_col = 'Customer nr.')

After that, we define a function in order to check discrepancies.

In [3]:
def compensation_test(customer_claim,sap):
    customer_claims_ids = customer_claim.groupby(customer_claim.index)\
    ['QTY to compensate','End stock/Sellin'].sum()
    sap_ids = sap.groupby(sap.index)['Processed quantity'].sum()
    
    not_compensated = [(sap_id,"not_compensated") for sap_id 
                            in sap_ids.index
                               if sap_id not in customer_claims_ids.index ]
    
    overcompensated =  [(customer_claims_id,'overcompensated') 
                            for customer_claims_id
                                in customer_claims_ids.index
                                    if customer_claims_id not in sap_ids.index]
    
    wrong_claims = [(customer_claim_id,'wrong_claim') for customer_claim_id
                    in customer_claims_ids.index
                        if (customer_claims_ids.loc[customer_claim_id]['QTY to compensate'] - sap_ids.loc[customer_claim_id] != 0)]
    
    suspicious_compensations =  not_compensated + overcompensated + wrong_claims
    return  suspicious_compensations

In [4]:
compensation_test(df_cc,df_sap)

[(271500357, 'not_compensated'), (272500650, 'wrong_claim')]

Two discrepancies:
- 271500357 doesn't exist on customer claim sheet.
- 272500650 Claimed one unit more than corresponding on SAP registry.

<img src="./ex2.JPG">

First of all I import csv files extracted from the excel sheet.

In [5]:
df_ac = pd.read_csv("./SONY_Exercise_2 - Amounts to be compensated.csv",index_col = 'Model name' )
df_bq = pd.read_csv("./SONY_Exercise_2 - Billing qtys.csv",index_col = 'Material')

I define a function in order to calculate the costs.

In [6]:
def calculate_total_amount(amount_compensated,billing_quantitys):
    
    billing_quantitys['amount'] = [ amount_compensated.loc[index]['Amount/PC (CZK)']*qty
                                   for (index,qty) 
                                       in zip(billing_quantitys.index,billing_quantitys['Bill qty'])]
    
    grouped_billing_quantitys = billing_quantitys.groupby(billing_quantitys.Name)['amount'].sum()
    
    return grouped_billing_quantitys,grouped_billing_quantitys.sum()

In [7]:
calculate_total_amount(df_ac,df_bq)

(Name
 AT COMPUTERS A.S.               8852
 Aaron shop s.r.o.              18620
 BER-KR SERVICES S.R.O.          4767
 CEWE COLOR a.s.                15820
 ELLEX GK S.R.O.                 5497
 Elvia pro, spol. s r.o.        16348
 MEGAPIXEL S.R.O.               20732
 Milan Skoda-FOTO               19556
 SWS A.S.                        8104
 VIDEO klinik, spol. s r. o.     1424
 Name: amount, dtype: int64, 119720)

Total costs is 119720

<img src="./ex3.JPG">

First of all I import csv file extracted from the excel sheet.

In [8]:
df_fy = pd.read_csv("./SONY_Exercise_3 - FY13 Sept-March DI.csv",index_col = 'Bill.Doc.')

I parse 'bill date' colum to datetime type.

In [9]:
df_fy['date'] = pd.to_datetime(df_fy['Bill. Date'])

now we calculate the amount of bills on 2013-12-20

In [10]:
ondate = df_fy.loc[df_fy['date'] == '2013-12-20']

In [11]:
ondate

Item       Material  MPG  Billed qty  CnTy     Amount  CondRecNo.  \
Bill.Doc.                                                                       
6876100217   260    SLTA58K.CEC  NaN           2  ZNET        130         NaN   
6876100218   120  DSCHX300B.CE3  NaN           2  ZNET         70         NaN   
6876100219    60   DSCHX50B.CE3  NaN           1  ZNET         60         NaN   
6876100220    70   DSCHX50S.CE3  NaN           2  ZNET         60         NaN   
6876100221   150    SLTA58K.CEC  NaN           1  ZNET        130         NaN   
6876100222   110  DSCRX100.CEE8  NaN           1  ZNET        163         NaN   
6876100223    50  DSCRX100.CEE8  NaN           1  ZNET        163         NaN   
6876100224    40  DSCWX200B.CE3  NaN           2  ZNET         50         NaN   
6876100225    60  DSCWX200B.CE3  NaN           2  ZNET         50         NaN   
6876100226   110  DSCWX200S.CE3  NaN           2  ZNET         50         NaN   
6876100227   160  DSCWX200S.CE3  NaN           1  ZNET         50         NaN   
6876100228    90   DSCHX50B.CE3  NaN           2  ZNET         60         NaN   
6876100229   120  DSCWX300B.CE3  NaN           1  ZNET         60         NaN   
6876100230   180  DSCWX300L.CE3  NaN           1  ZNET         60         NaN   
6876100231   130  DSCHX300B.CE3  NaN           7  ZNET         70         NaN   
6876100232   280  DSCHX300B.CE3  NaN           1  ZNET         70         NaN   
6876100233    10    SLTA58K.CEC  NaN           1  ZNET        130         NaN   
6876100234    40    SLTA58K.CEC  NaN           1  ZNET        130         NaN   
6876100235    10  DSCHX300B.CE3  NaN           2  ZNET         70         NaN   
6876100236    30   DSCHX50B.CE3  NaN           1  ZNET         60         NaN   
6876100237    40  DSCWX300B.CE3  NaN           1  ZNET         60         NaN   
6876100239    10    SLTA58K.CEC  NaN           3  ZNET        130         NaN   
6876100240   150  DSCWX200B.CE3  NaN           1  ZNET         50         NaN   
6876100241    10  DSCHX300B.CE3  NaN           3  ZNET         70         NaN   
6876100242    40  DSCHX300B.CE3  NaN           2  ZNET         70         NaN   
6876100243    50   DSCHX50B.CE3  NaN           2  ZNET         60         NaN   
6876100244    60  DSCWX300B.CE3  NaN           1  ZNET         60         NaN   
6876100245    20  DSCWX300L.CE3  NaN           1  ZNET         60         NaN   
6876100246    70  DSCWX300W.CE3  NaN           1  ZNET         60         NaN   
6876100247   170    SLTA58K.CEC  NaN           2  ZNET        130         NaN   
...          ...            ...  ...         ...   ...        ...         ...   
6876100688    30  DSCWX200S.CE3  NaN           1  ZNET         50         NaN   
6876100689    10   DSCHX50B.CE3  NaN           3  ZNET         60         NaN   
6876100690    20   DSCHX50B.CE3  NaN           1  ZNET         60         NaN   
6876100691    10   DSCHX50S.CE3  NaN           2  ZNET         60         NaN   
6876100692   110  DSCWX300B.CE3  NaN           2  ZNET         60         NaN   
6876100693   130  DSCWX300L.CE3  NaN           1  ZNET         60         NaN   
6876100694    40  DSCHX300B.CE3  NaN           1  ZNET         70         NaN   
6876100695   150  DSCHX300B.CE3  NaN           1  ZNET         70         NaN   
6876100696    10    SLTA58K.CEC  NaN           2  ZNET        130         NaN   
6876100697    60    SLTA58K.CEC  NaN           1  ZNET        130         NaN   
6876100698    50  DSCRX100.CEE8  NaN           1  ZNET        163         NaN   
6876100699    90  DSCWX300B.CE3  NaN           1  ZNET         60         NaN   
6876100700   110  DSCWX300L.CE3  NaN           2  ZNET         60         NaN   
6876100701   250  DSCHX300B.CE3  NaN           1  ZNET         70         NaN   
6876100702    50    SLTA58K.CEC  NaN           1  ZNET        130         NaN   
6876100703    50  DSCRX100.CEE8  NaN           1  ZNET        163         NaN   
6876100704    10   DSCHX50B.CE3  NaN       

481 invoices were created on 20th of December of 2013

In [12]:
krakow_invoices = df_fy.loc[df_fy['Name 1'] == 'Neonet S.A. Krakow']

In [13]:
krakow_invoices

Item       Material  MPG  Billed qty  CnTy     Amount  CondRecNo.  \
Bill.Doc.                                                                       
6876107368    60  ILCE3000KB.CE  NaN          59  ZNET        210         NaN   
6876112253    10   DSCW710B.CE3  NaN           5  ZNET         10         NaN   
6876112253    20   DSCW710S.CE3  NaN           5  ZNET         10         NaN   
6876112254   131   DSCW710B.CE3  NaN          70  ZNET         10         NaN   
6876112255    20   DSCW710B.CE3  NaN          20  ZNET         10         NaN   
6876112256    20   DSCW710B.CE3  NaN          10  ZNET         10         NaN   
6876112257    20   DSCW710B.CE3  NaN          10  ZNET         10         NaN   
6876112258    30   DSCW710S.CE3  NaN          10  ZNET         10         NaN   
6876112258    50   DSCW730G.CE3  NaN           1  ZNET         12         NaN   
6876112259    30   DSCW710S.CE3  NaN          20  ZNET         10         NaN   
6876112260    50   DSCW710S.CE3  NaN          20  ZNET         10         NaN   
6876112261    40   DSCW710S.CE3  NaN          10  ZNET         10         NaN   
6876112262    41   DSCW710S.CE3  NaN           8  ZNET         10         NaN   
6876112263    70   DSCH200B.CE3  NaN           5  ZNET         50         NaN   
6876112264    10   DSCH200B.CE3  NaN          40  ZNET         50         NaN   
6876112265    11   DSCH200B.CE3  NaN          21  ZNET         50         NaN   
6876112266    10  DSCHX300B.CE3  NaN          10  ZNET        140         NaN   

            Agreement   Cond.value SOff.    ...       Product hierarchy  \
Bill.Doc.                                   ...                           
6876107368        NaN    12.390,00  PL00    ...      SCPCDIMAMCNEXKNEXK   
6876112253        NaN           50  PL00    ...      SCPCDIMDSCCYBRCYBS   
6876112253        NaN           50  PL00    ...      SCPCDIMDSCCYBRCYBS   
6876112254        NaN          700  PL00    ...      SCPCDIMDSCCYBRCYBS   
6876112255        NaN          200  PL00    ...      SCPCDIMDSCCYBRCYBS   
6876112256        NaN          100  PL00    ...      SCPCDIMDSCCYBRCYBS   
6876112257        NaN          100  PL00    ...      SCPCDIMDSCCYBRCYBS   
6876112258        NaN          100  PL00    ...      SCPCDIMDSCCYBRCYBS   
6876112258        NaN           12  PL00    ...      SCPCDIMDSCCYBRCYBS   
6876112259        NaN          200  PL00    ...      SCPCDIMDSCCYBRCYBS   
6876112260        NaN          200  PL00    ...      SCPCDIMDSCCYBRCYBS   
6876112261        NaN          100  PL00    ...      SCPCDIMDSCCYBRCYBS   
6876112262        NaN           80  PL00    ...      SCPCDIMDSCCYBRCYBS   
6876112263        NaN          250  PL00    ...      SCPCDIMDSCCYBRCYBS   
6876112264        NaN     2.000,00  PL00    ...      SCPCDIMDSCCYBRCYBS   
6876112265        NaN     1.050,00  PL00    ...      SCPCDIMDSCCYBRCYBS   
6876112266        NaN     1.400,00  PL00    ...      SCPCDIMDSCCYBRCYBS   

             MG3* MG4*   Created OrdRs Sales Doc.   Doc. Date  \
Bill.Doc.                                                       
6876107368  100.0  NaN  HUPERJEK   C03   67999956  11.02.2014   
6876112253  100.0  NaN  HUPERJEK   C03   68063693  13.03.2014   
6876112253  100.0  NaN  HUPERJEK   C03   68063693  13.03.2014   
6876112254  100.0  NaN  HUPERJEK   C03   68063694  13.03.2014   
6876112255  100.0  NaN  HUPERJEK   C03   68063695  13.03.2014   
6876112256  100.0  NaN  HUPERJEK   C03   68063696  13.03.2014   
6876112257  100.0  NaN  HUPERJEK   C03   68063697  13.03.2014   
6876112258  100.0  NaN  HUPERJEK   C03   68063698  13.03.2014   
6876112258  100.0  NaN  HUPERJEK   C03   68063698  13.03.2014   
6876112259  100.0  NaN  HUPERJEK   C03   68063699  13.03.2014   
6876112260  100.0  NaN  HUPERJEK   C03   68063701  13.03.2014   
6876112261  100.0  NaN  HUPERJEK   C03   68063702  13.03.2014   
6876112262  100.0  NaN  HUPERJEK   C03   68063703  13.03.2014   
6876112263  100.0  NaN  HUPERJEK   C03   68063705  13.03.2014   
687611

17 invoices were created for Neonet S.A. Krakow

Now we select Billed quantities greather than 60

In [14]:
filtered = df_fy.loc[df_fy['Billed qty'] > 60][['Billed qty','  Cond.value']]

Cond.value column have an invalid format for numbers, so we must transform them.

In [15]:
filtered['  Cond.value'] = [value.replace('.','')for value in filtered['  Cond.value']]
filtered['  Cond.value'] = [value.replace(',','.')for value in filtered['  Cond.value']]
filtered['  Cond.value'] = [float(value) for value in filtered['  Cond.value']]

In [16]:
filtered


Billed qty    Cond.value
Bill.Doc.                           
6876083708          70       3500.00
6876083808          92       9752.00
6876086991         123       5043.00
6876086992          77       9240.00
6876086998          94      11280.00
6876087834          87      10440.00
6876107370          70      14700.00
6876107730          71      16330.00
6876107739          67      15410.00
6876110028          90       4500.00
6876110029         160       8000.00
6876110030         150       7500.00
6876110033         115       5750.00
6876110035          85       4250.00
6876110036          75       3750.00
6876110037         219      30660.00
6876110038         131      18340.00
6876110041         176      24640.00
6876110100         104      14560.00
6876110105         103       1030.00
6876110107          82        984.00
6876110272         109      15260.00
6876110277         152      21280.00
6876110394          66       9240.00
6876112254          70        700.00
6876112268          90        900.00
6876112269          97        970.00
6876112463          65       8125.00
6876113670          88      22000.00
6876113682          78       3900.00
6876113686          94      13160.00
6876113687         180       7200.00
6876113688         150       6000.00
6876113689         157       7850.00
6876113695         105       5250.00
6876113703          94       4700.00
6876113896         119      16660.00
6876113897         156      21840.00
6876113968          98       4900.00
6876116906          67        722.26
6876116909         110       1692.90
6876116915         132       1422.96
6876116917         120       1293.60
6876116918         139       1498.42
6876116919          86        927.08
6876116932         176       2708.64
6876116938         140       2154.60
6876116940         126       3451.14
6876116943          80       1231.20

In [17]:
filtered.sum()

Billed qty        5385.0
  Cond.value    406696.8
dtype: float64

- Total amount of bills is  5385.0
- Total amount of conditional value is 406696.8